apel/db/records/summary.py

```diff
@@ -225,3 +227,53 @@ def get_ur(self):
         # We don't want the XML declaration, because the whole XML
         # document will be assembled by another part of the program.
         return doc.documentElement.toxml()
+
+
+class SummaryRecord04(SummaryRecord):
+    """Class to represent a summary record using the 0.4 message format
+
+    It differs from SummaryRecord by lacking a separate ServiceLevelType field
+    in the message fields, as this is extracted from the associative array
+    in ServiceLevel before putting into the database.
+    """
+
+    def __init__(self):
+        """Populate fields required to load the message."""
+
+        Record.__init__(self)
+
+        # Fields which are required by the message format.
+        self._mandatory_fields = ["Site", "Month", "Year", "WallDuration",
+                                  "CpuDuration", "NumberOfJobs"]
+
+        # This list allows us to specify the order of lines when we construct records.
+        # It differs from SummaryRecord by lacking a separate ServiceLevelType field
+        # as this is included in the dict for ServiceLevel.
+        self._msg_fields = [
+            "Site", "Month", "Year", "GlobalUserName", "VO", "VOGroup", "VORole", "SubmitHost",
+            "InfrastructureType", "ServiceLevel", "NodeCount", "Processors",
+            "EarliestEndTime", "LatestEndTime", "WallDuration", "CpuDuration", "NumberOfJobs"
+        ]
+
+        # This list specifies the information that goes in the database.
+        self._db_fields = [
+            "Site", "Month", "Year", "GlobalUserName", "VO", "VOGroup", "VORole", "SubmitHost",
+            "InfrastructureType", "ServiceLevelType", "ServiceLevel", "NodeCount", "Processors",
+            "EarliestEndTime", "LatestEndTime", "WallDuration", "CpuDuration", "NumberOfJobs"
+        ]
+
+        self._ignored_fields = ["UpdateTime"]
+
+        # All allowed fields. We use _db_fields as that is a superset of _msg_fields.
+        self._all_fields = self._db_fields
+
+        # Fields which will have an integer stored in them
+        self._int_fields = ["Month", "Year", "NodeCount", "Processors",
+                            "WallDuration", "CpuDuration", "NumberOfJobs"]
+
+        self._float_fields = ["ServiceLevel"]
+
+        self._datetime_fields = ["EarliestEndTime", "LatestEndTime"]
+
+        # Fields which should contain associative arrays in the message
+        self._dict_fields = ["ServiceLevel"]
```

Comment:
Again, not an expert, should ServiceLevel be considered in two different fields for two different datatypes?  Is this one of those things where SL is going to be a float in one case and a dict in others?